### PR TITLE
Implement Collection Operation to Relabelling Identifiers

### PR DESCRIPTION
**Pull request by jmchilton**
... against supplied labels from an input dataset of type txt.

I think it needs to be touched up but the basic operation seems to work so far. I think what remains to be done is:

 - [x] Validate uniqueness of identifiers and provide nice messages if they are not unique.
 - [x] Validate that at least the required number of lines are present in the file and provide a nice message if not.
 - [x] Add strict mode to ensure exactly the correct number of lines is added.
 - [x] Find where validation of identifiers happens in the API and apply same validation here - try not to let unsafe identifiers be created.
 - [x] Consider more advanced modes - selecting a column, apply a regex replace, pick two columns for nested lists, etc.... None of this may need to be needed in the first iteration.
 - [x] ~Consider another mode where a collection is labelled against an existing collection - should that be a separate tool of the same tool.~

A collection operation variant of the tool from @pjbriggs at https://github.com/pjbriggs/Amplicon_analysis-galaxy/blob/77340d8bb2470a646deba4933625413fc70985d1/relabel_samples.xml.

### CODE REVIEW

**Comment by mvdbeek** [comment]
> Consider more advanced modes - selecting a column, apply a regex replace, pick two columns for nested lists, etc.... None of this may need to be needed in the first iteration.
> Consider another mode where a collection is labelled against an existing collection - should that be a separate tool of the same tool.

The last point could be solved with a tool that just dumps the element identifiers (that wouldn't necessarily need to be shipped with galaxy), and then you're free to use all the fantastic column/ regex tools in galaxy land to build a 2 column list (if we make this tool accept a 2 column list ...) .

**Comment by pjbriggs** [comment]
@jmchilton thanks again for making this first implementation, it looks really useful. I've patched our local Galaxy and the tool works as advertised for me.

A couple of observations/suggestions from initial usage:

- Could  you allow the user to also specify the new name for the relabelled collection? (There doesn't seem to be a way to rename a collection once it's been created.)
- It's potentially quite easy to accidentally mislabel collection items with the single column list, as it relies on the user correctly ordering the new labels in the input file. The option of using a 2 column list specifying old and new identifiers would help to mitigate this (as it's easier to see what will map to what).

HTH.  Thanks again!

**Comment by nsoranzo** [comment]
> Could you allow the user to also specify the new name for the relabelled collection? (There doesn't seem to be a way to rename a collection once it's been created.)

If you click on the collection in the history panel, it will show the list of the collection datasets. There you can rename the collection by clicking on its name.

**Comment by pjbriggs** [comment]
@nsoranzo thanks! I never noticed that before (I was looking for something like the "pencil" icon on regular datasets).

**Comment by mvdbeek** [comment]
@pjbriggs @jmchilton I've opened a PR with a simple implementation of a 2-column list rename [here](https://github.com/jmchilton/galaxy/pull/54)

File: lib/galaxy/tools/relabel_from_file.xml [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                <collection type="list:paired">
+                    <element name="i1">
+                        <collection type="paired">
+                            <element name="forward" value="simple_line.txt" />
+                            <element name="reverse" value="simple_line_alternative.txt" />
+                        </collection>
+                    </element>
+                </collection>
+            </param>
+            <param name="labels" value="new_labels_2.txt" ftype="txt" />
```
Comment (mvdbeek) [inline lib/galaxy/tools/relabel_from_file.xml]: ftype needs to be tabular, that's why the test failed :(.

**Comment by jmchilton** [comment]
@mvdbeek I reworked that different modality to be an explicit user choice - I hope that is okay (see 0057dc616319ae78031926d4c24c1898b24987a9). I'd rather not have such a large change in behavior depend on a subtle datatype difference if that is okay.

**Comment by jmchilton** [comment]
A couple more fixes rebased into 09bcdd9 so I am pulling this out of WIP.

@mvdbeek Your mapping approach is pretty cool and it makes me think the two missing collections - "group by" and "filter" could easily be implemented by inspecting a table this way. I'd say someday we would still want a "filter by expression" and "group by expression" tools - just like we'd like to have "relabel by expression" - but these should exist beside "relabel by file", "group by file", and "filter by file" tools. I'll try to get something ready for 17.05 - it might be a stretch though.

File: lib/galaxy/tools/__init__.py [hunk 2]
```diff
@@ -2506,11 +2506,56 @@ def add_elements(collection, prefix=""):
...
+        strict = string_as_bool(incoming["how"]["strict"])
+        new_elements = odict()
+        new_labels_path = new_labels_dataset_assoc.file_name
+        new_labels = open(new_labels_path, "r").readlines(1024 * 1000000)
+        if strict and len(hdca.collection.elements) != len(new_labels):
+            raise Exception("Relabel mapping file contains incorrect number of identifiers")
+        if how_type == "tabular":
+            # We have a tabular file, where the first column is an existing element identifier,
+            # and the second column is the new element identifier.
+            source_new_label = (line.strip().split('\t') for line in new_labels)
```
Comment (mvdbeek) [inline lib/galaxy/tools/__init__.py:2535]: Maybe we should use `strip('\r\n')` here? See the discussion here: https://github.com/galaxyproject/tools-iuc/pull/1233#discussion_r109772842

**Comment by mvdbeek** [comment]
> I reworked that different modality to be an explicit user choice - I hope that is okay (see 0057dc6). I'd rather not have such a large change in behavior depend on a subtle datatype difference if that is okay.

That's better, I agree.

> I'll try to get something ready for 17.05 - it might be a stretch though.

Is that for the  "filter by expression" and "group by expression" tools ? I think this PR looks good to me, from your bullet-points I think 1,2 and 3 are addressed, right?

**Comment by jmchilton** [comment]
> Is that for the "filter by expression" and "group by expression" tools ?

No - I meant I hope to open the "from file" variants this week. The more I think about it - the more I like them. The expression stuff will come in a future release.

> I think this PR looks good to me, from your bullet-points I think 1,2 and 3 are addressed, right?

Now those bullet points are indeed - I added one more test case and fix to cover duplicate identifiers in particular. I think the only bullet not addressed now is "Consider another mode where a collection is labelled against an existing collection - should that be a separate tool of the same tool." Which I am happy to push off for now.
